In [2]:
import pandas as pd
import numpy as np
import ast


### Importando los datos

In [3]:
filas=[]
with open('steam_games.json') as a:
    for linea in a.readlines():
        filas.append(ast.literal_eval(linea))
juegos=pd.DataFrame(filas)

In [4]:
juegos

publisher                                             genres  \
0             Kotoshiro      [Action, Casual, Indie, Simulation, Strategy]   
1      Making Fun, Inc.               [Free to Play, Indie, RPG, Strategy]   
2          Poolians.com  [Casual, Free to Play, Indie, Simulation, Sports]   
3                  彼岸领域                        [Action, Adventure, Casual]   
4                   NaN                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games              [Casual, Indie, Simulation, Strategy]   
32131            Sacada                          [Casual, Indie, Strategy]   
32132      Laush Studio                        [Indie, Racing, Simulation]   
32133          SIXNAILS                                    [Casual, Indie]   
32134               NaN                                                NaN   

                       app_name                     title  \
0           Lost Summoner Kitty       Lost Summoner Kitty   
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                       NaN   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url release_date  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07   
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
...                                                  ...          ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  discount_price  \
0          [Strategy, Action, Indie, Casual, Simulation]            4.49   
1      [Free to Play, Strategy, Indie, RPG, Card Game...             NaN   
2      [Free to Play, Simulation, Sports, Casual, Ind...             NaN   
3                            [Action, Adventure, Casual]            0.83   
4                        [Action, Indie, Casual, Sports]            1.79   
...                                                  ...             ...   
32130              [Strategy, Indie, Casual, Simulation]            1.49   
32131                          [Strategy, Indie, Casual]            4.24   
32132                        [Indie, Simulation, Racing]            1.39   
32133  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...             NaN   
32134  [Early Access, Adventure, Indie, Action, Simul...             NaN   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://

## Limpieza de los datos (sacar valores NaN, nulos, explorar si hay similitudes entre columnas, etc)

### 1- Columna 'publisher'

In [5]:
#########
# cambiamos los datos a tipo str, elminamos NaN y cambaimos por la cadena de texto 'no_info':
#########
juegos['publisher']=juegos['publisher'].astype('str')
juegos['publisher'].replace(to_replace='nan', value='no_info', inplace=True)
juegos['publisher'].replace(to_replace='N/A', value='no_info', inplace=True)
juegos['publisher'].replace(to_replace='NA', value='no_info', inplace=True)
#NA

In [6]:
juegos

publisher                                             genres  \
0             Kotoshiro      [Action, Casual, Indie, Simulation, Strategy]   
1      Making Fun, Inc.               [Free to Play, Indie, RPG, Strategy]   
2          Poolians.com  [Casual, Free to Play, Indie, Simulation, Sports]   
3                  彼岸领域                        [Action, Adventure, Casual]   
4               no_info                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games              [Casual, Indie, Simulation, Strategy]   
32131            Sacada                          [Casual, Indie, Strategy]   
32132      Laush Studio                        [Indie, Racing, Simulation]   
32133          SIXNAILS                                    [Casual, Indie]   
32134           no_info                                                NaN   

                       app_name                     title  \
0           Lost Summoner Kitty       Lost Summoner Kitty   
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                       NaN   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url release_date  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07   
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
...                                                  ...          ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  discount_price  \
0          [Strategy, Action, Indie, Casual, Simulation]            4.49   
1      [Free to Play, Strategy, Indie, RPG, Card Game...             NaN   
2      [Free to Play, Simulation, Sports, Casual, Ind...             NaN   
3                            [Action, Adventure, Casual]            0.83   
4                        [Action, Indie, Casual, Sports]            1.79   
...                                                  ...             ...   
32130              [Strategy, Indie, Casual, Simulation]            1.49   
32131                          [Strategy, Indie, Casual]            4.24   
32132                        [Indie, Simulation, Racing]            1.39   
32133  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...             NaN   
32134  [Early Access, Adventure, Indie, Action, Simul...             NaN   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://

### 2- Columna 'genres'

Cada registro de este campo tiene una lista. He intentado cambiar a tipo string, pero en esta limpieza no he podido avanzar. Lo dejo como esta sin cambios.

In [7]:
juegos.loc[juegos['genres'].isna()== True]


publisher genres                                    app_name  \
4       no_info    NaN                               Log Challenge   
11      no_info    NaN                        Icarus Six Sixty Six   
19      no_info    NaN                               After Life VR   
20      no_info    NaN                                  Kitty Hawk   
22      no_info    NaN                                  Mortars VR   
...         ...    ...                                         ...   
32118   no_info    NaN   Robotpencil Presents: Exercise: Brushwork   
32119   no_info    NaN  Robotpencil Presents: Creative Composition   
32120   no_info    NaN                            The Gamble House   
32121   no_info    NaN    Kalen Chock Presents: 2017 Free Tutorial   
32134   no_info    NaN                                 Maze Run VR   

                                            title  \
4                                             NaN   
11                                            NaN   
19                                            NaN   
20                                            NaN   
22                                            NaN   
...                                           ...   
32118   Robotpencil Presents: Exercise: Brushwork   
32119  Robotpencil Presents: Creative Composition   
32120                            The Gamble House   
32121    Kalen Chock Presents: 2017 Free Tutorial   
32134                                         NaN   

                                                     url release_date  \
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
11     http://store.steampowered.com/app/724910/Icaru...          NaN   
19     http://store.steampowered.com/app/772590/After...          NaN   
20     http://store.steampowered.com/app/640250/Kitty...          NaN   
22     http://store.steampowered.com/app/711440/Morta...          NaN   
...                                                  ...          ...   
32118  http://store.steampowered.com/app/775640/Robot...   2018-01-03   
32119  http://store.steampowered.com/app/777930/Robot...   2018-01-03   
32120  http://store.steampowered.com/app/775370/The_G...   2016-11-19   
32121  http://store.steampowered.com/app/777950/Kalen...   2018-01-03   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  discount_price  \
4                        [Action, Indie, Casual, Sports]            1.79   
11                                              [Casual]             NaN   
19                             [Early Access, Indie, VR]             NaN   
20      [Early Access, Action, Adventure, Indie, Casual]            1.97   
22     [Early Access, Strategy, Action, Indie, Casual...            0.89   
...                                                  ...             ...   
32118                  [Design & Illustration, Tutorial]             NaN   
32119                  [Design & Illustration, Tutorial]             NaN   
32120                                            [Movie]             NaN   
32121                  [Design & Illustration, Tutorial]             NaN   
32134  [Early Access, Adventure, Indie, Action, Simul...             NaN   

                                             reviews_url  \
4      http://steamcommunity.com/app/773570/reviews/?...   
11     http://steamcommunity.com/app/724910/reviews/?...   
19     http://steamcommunity.com/app/772590/reviews/?...   
20     http://steamcommunity.com/app/640250/reviews/?...   
22     http://steamcommunity.com/app/711440/reviews/?...   
...                                                  ...   
32118  http://steamcommunity.com/app/775640/reviews/?...   
32119  http://steamcommunity.com/app/777930/reviews/?...   
32120  http://steamcommunity.com/app/775370/reviews/?...   
32121  http://steamcommunity.com/app/777950/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                 

In [8]:
#############################
#con esto quise hacer que la lista deje ser una lista y sea una cadena de caracteres, pero no lo logré.
#############################

#juegos['genres']=juegos['genres'].astype('str')
#juegos['genres'].replace(to_replace='NaN', value='no_info', inplace=True)
#juegos['genres'].replace(to_replace='', value='no_info', inplace=True)
#juegos['genres'].replace(to_replace='[', value='', inplace=True)
#juegos['genres'].replace(to_replace=']', value='', inplace=True) 

### 3- Columna 'app_name'

In [9]:
#############
# Con esta columna hago lo mismo que con la columna 'publisher'; paso a tipo de dato str y cambio valores nan por 'no_info'. Esto mismo lo hago en los siguientes campos.
#############
juegos['app_name']=juegos['app_name'].astype('str')
juegos['app_name'].replace(to_replace= 'nan' , value='no_info', inplace=True)


Dada la similitud entre 'app_name' y 'title', se me ocurre remplazar los registros 'no_info' de app_name por los de 'title'. Primero filtro los 'no_info':

In [10]:

juegos.loc[juegos['app_name'] == 'no_info']



publisher           genres app_name title  \
74     no_info              NaN  no_info   NaN   
2580   no_info  [Action, Indie]  no_info   NaN   

                                              url release_date  \
74                 http://store.steampowered.com/          NaN   
2580  http://store.steampowered.com/app/317160/_/   2014-08-26   

                 tags  discount_price  \
74                NaN           14.99   
2580  [Action, Indie]             NaN   

                                            reviews_url  \
74                                                  NaN   
2580  http://steamcommunity.com/app/317160/reviews/?...   

                           specs  price  early_access      id developer  \
74                           NaN  19.99         False     NaN       NaN   
2580  [Single-player, Game demo]    NaN         False  317160       NaN   

     sentiment metascore  
74         NaN       NaN  
2580  Positive       NaN

#### Conclusion: 
Los registros 'no_info' de 'app name', son 'NaN' en 'title'.

### 4- Columna 'title'

In [11]:
juegos['title']=juegos['title'].astype('str')
juegos['title'].replace(to_replace= 'nan' , value='no_info', inplace=True)

In [12]:
#juegos.loc[juegos['app_name'] != juegos['title'] ]
juegos

publisher                                             genres  \
0             Kotoshiro      [Action, Casual, Indie, Simulation, Strategy]   
1      Making Fun, Inc.               [Free to Play, Indie, RPG, Strategy]   
2          Poolians.com  [Casual, Free to Play, Indie, Simulation, Sports]   
3                  彼岸领域                        [Action, Adventure, Casual]   
4               no_info                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games              [Casual, Indie, Simulation, Strategy]   
32131            Sacada                          [Casual, Indie, Strategy]   
32132      Laush Studio                        [Indie, Racing, Simulation]   
32133          SIXNAILS                                    [Casual, Indie]   
32134           no_info                                                NaN   

                       app_name                     title  \
0           Lost Summoner Kitty       Lost Summoner Kitty   
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                   no_info   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                   no_info   

                                                     url release_date  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07   
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
...                                                  ...          ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  discount_price  \
0          [Strategy, Action, Indie, Casual, Simulation]            4.49   
1      [Free to Play, Strategy, Indie, RPG, Card Game...             NaN   
2      [Free to Play, Simulation, Sports, Casual, Ind...             NaN   
3                            [Action, Adventure, Casual]            0.83   
4                        [Action, Indie, Casual, Sports]            1.79   
...                                                  ...             ...   
32130              [Strategy, Indie, Casual, Simulation]            1.49   
32131                          [Strategy, Indie, Casual]            4.24   
32132                        [Indie, Simulation, Racing]            1.39   
32133  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...             NaN   
32134  [Early Access, Adventure, Indie, Action, Simul...             NaN   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://

Ahora buscaré en que se diferencian 'title' de 'app_name', cuando los valores de los registros son distintos.


In [13]:
juegos.loc[(juegos['title']!= 'no_info') & (juegos['title'] != juegos['app_name'])]
#juegos.loc[(juegos['title'].isna() != True) & (juegos['title'] != juegos['app_name'])]

publisher               genres  \
80      Telltale Games  [Action, Adventure]   
83      Telltale Games  [Action, Adventure]   
109    Electronic Arts           [Strategy]   
182            Ubisoft           [Strategy]   
184            Ubisoft           [Strategy]   
...                ...                  ...   
31871   Telltale Games  [Action, Adventure]   
31872   Telltale Games  [Action, Adventure]   
31873   Telltale Games  [Action, Adventure]   
31898   Strategy First   [Action, Strategy]   
32052          Ubisoft        [Action, RPG]   

                                            app_name  \
80                      Sam & Max 101: Culture Shock   
83                  Sam & Max 102: Situation: Comedy   
109                   Command & Conquer: Red Alert 3   
182       Heroes of Might & Magic V: Hammers of Fate   
184    Heroes of Might & Magic V: Tribes of the East   
...                                              ...   
31871                     Sam & Max 105: Reality 2.0   
31872           Sam & Max 104: Abe Lincoln Must Die!   
31873         Sam & Max 106: Bright Side of the Moon   
31898           Making History: The Calm & the Storm   
32052                  Dark Messiah of Might & Magic   

                                                   title  \
80                      Sam &amp; Max 101: Culture Shock   
83                  Sam &amp; Max 102: Situation: Comedy   
109                   Command &amp; Conquer: Red Alert 3   
182       Heroes of Might &amp; Magic V: Hammers of Fate   
184    Heroes of Might &amp; Magic V: Tribes of the East   
...                                                  ...   
31871                     Sam &amp; Max 105: Reality 2.0   
31872           Sam &amp; Max 104: Abe Lincoln Must Die!   
31873         Sam &amp; Max 106: Bright Side of the Moon   
31898           Making History: The Calm &amp; the Storm   
32052                  Dark Messiah of Might &amp; Magic   

                                                     url release_date  \
80     http://store.steampowered.com/app/8200/Sam__Ma...   2006-10-17   
83     http://store.steampowered.com/app/8210/Sam__Ma...   2006-12-20   
109    http://store.steampowered.com/app/17480/Comman...   2008-10-28   
182    http://store.steampowered.com/app/15380/Heroes...   2006-11-14   
184    http://store.steampowered.com/app/15370/Heroes...   2007-10-16   
...                                                  ...          ...   
31871  http://store.steampowered.com/app/8240/Sam__Ma...   2007-03-29   
31872  http://store.steampowered.com/app/8230/Sam__Ma...   2007-02-22   
31873  http://store.steampowered.com/app/8250/Sam__Ma...   2007-04-26   
31898  http://store.steampowered.com/app/6250/Making_...   2007-03-13   
32052  http://store.steampowered.com/app/2100/Dark_Me...   2006-10-25   

                                                    tags  discount_price  \
80     [Point & Click, Comedy, Adventure, Detective, ...             NaN   
83                                   [Adventure, Action]             NaN   
109    [Strategy, RTS, Base Building, Multiplayer, Co...             NaN   
182    [Strategy, Turn-Based Strategy, Turn-Based, Fa...             NaN   
184    [Strategy, Turn-Based Strategy, Turn-Based, Fa...             NaN   
...                                                  ...             ...   
31871                                [Adventure, Action]             NaN   
31872  [Adventure, Action, Point & Click, Free to Pla...             NaN   
31873                                [Adventure, Action]             NaN   
31898  [Strategy, Action, Turn-Based Strategy, Grand ...             NaN   
32052  [RPG, Action, First-Person, Fantasy, Adventure...             NaN   

                                             reviews_url  \
80     http://steamcommunity.com/app/8200/reviews/?br...   
83     http://steamcommunity.com/app/8210/reviews/?br...   
109    http://steamcommunity.com/app/17480/reviews/?b...   
182    http://steamcommu

Como se ve en la tabla anterior, las diferencias vienen de que en la columna 'title' despues del & le sigue un 'amp;'. Parece ser un error de compilacion o de formatos. Para verificarlo, vemos cuantas registros de la columna 'title' contienen un '&':

In [14]:

juegos.loc[juegos['title'].str.contains('&', regex=True)]

publisher               genres  \
80      Telltale Games  [Action, Adventure]   
83      Telltale Games  [Action, Adventure]   
109    Electronic Arts           [Strategy]   
182            Ubisoft           [Strategy]   
184            Ubisoft           [Strategy]   
...                ...                  ...   
31871   Telltale Games  [Action, Adventure]   
31872   Telltale Games  [Action, Adventure]   
31873   Telltale Games  [Action, Adventure]   
31898   Strategy First   [Action, Strategy]   
32052          Ubisoft        [Action, RPG]   

                                            app_name  \
80                      Sam & Max 101: Culture Shock   
83                  Sam & Max 102: Situation: Comedy   
109                   Command & Conquer: Red Alert 3   
182       Heroes of Might & Magic V: Hammers of Fate   
184    Heroes of Might & Magic V: Tribes of the East   
...                                              ...   
31871                     Sam & Max 105: Reality 2.0   
31872           Sam & Max 104: Abe Lincoln Must Die!   
31873         Sam & Max 106: Bright Side of the Moon   
31898           Making History: The Calm & the Storm   
32052                  Dark Messiah of Might & Magic   

                                                   title  \
80                      Sam &amp; Max 101: Culture Shock   
83                  Sam &amp; Max 102: Situation: Comedy   
109                   Command &amp; Conquer: Red Alert 3   
182       Heroes of Might &amp; Magic V: Hammers of Fate   
184    Heroes of Might &amp; Magic V: Tribes of the East   
...                                                  ...   
31871                     Sam &amp; Max 105: Reality 2.0   
31872           Sam &amp; Max 104: Abe Lincoln Must Die!   
31873         Sam &amp; Max 106: Bright Side of the Moon   
31898           Making History: The Calm &amp; the Storm   
32052                  Dark Messiah of Might &amp; Magic   

                                                     url release_date  \
80     http://store.steampowered.com/app/8200/Sam__Ma...   2006-10-17   
83     http://store.steampowered.com/app/8210/Sam__Ma...   2006-12-20   
109    http://store.steampowered.com/app/17480/Comman...   2008-10-28   
182    http://store.steampowered.com/app/15380/Heroes...   2006-11-14   
184    http://store.steampowered.com/app/15370/Heroes...   2007-10-16   
...                                                  ...          ...   
31871  http://store.steampowered.com/app/8240/Sam__Ma...   2007-03-29   
31872  http://store.steampowered.com/app/8230/Sam__Ma...   2007-02-22   
31873  http://store.steampowered.com/app/8250/Sam__Ma...   2007-04-26   
31898  http://store.steampowered.com/app/6250/Making_...   2007-03-13   
32052  http://store.steampowered.com/app/2100/Dark_Me...   2006-10-25   

                                                    tags  discount_price  \
80     [Point & Click, Comedy, Adventure, Detective, ...             NaN   
83                                   [Adventure, Action]             NaN   
109    [Strategy, RTS, Base Building, Multiplayer, Co...             NaN   
182    [Strategy, Turn-Based Strategy, Turn-Based, Fa...             NaN   
184    [Strategy, Turn-Based Strategy, Turn-Based, Fa...             NaN   
...                                                  ...             ...   
31871                                [Adventure, Action]             NaN   
31872  [Adventure, Action, Point & Click, Free to Pla...             NaN   
31873                                [Adventure, Action]             NaN   
31898  [Strategy, Action, Turn-Based Strategy, Grand ...             NaN   
32052  [RPG, Action, First-Person, Fantasy, Adventure...             NaN   

                                             reviews_url  \
80     http://steamcommunity.com/app/8200/reviews/?br...   
83     http://steamcommunity.com/app/8210/reviews/?br...   
109    http://steamcommunity.com/app/17480/reviews/?b...   
182    http://steamcommu

### 4- Columna 'title'

**Se elimina la columna**.
Puede verse que solo en 3 registros, la diferencia entre 'title' y 'app_name' no es por un NaN, ni por el error del '&amp'.  

Se ha demostrado que la columna no aporta informacion y por eso se elimina.


In [15]:
juegos=juegos.drop(['title'], axis=1)

In [16]:
pd.isnull(juegos).sum()

publisher             0
genres             3283
app_name              0
url                   0
release_date       2067
tags                163
discount_price    31910
reviews_url           2
specs               670
price              1377
early_access          0
id                    2
developer          3299
sentiment          7182
metascore         29458
dtype: int64

### 5- Columna 'realease_date'

El procedimiento de limpieza de esta columna es algo distinto a las otras. Primero cambio de tipo de dato, luego cambio algunos de los valores que no son del tipo año-mes-dia como por ejemplo 2020-10-23 por un valor de fecha 1900-01-01. Algunos otros como por ejemplo 'Jun 2009' los remplazo por '2009-06-01'.

In [17]:
juegos['release_date']=juegos['release_date'].astype(str)
juegos['release_date'].replace(to_replace='0000-01-01', value= '1900-01-01', inplace=True)
juegos['release_date'].replace(to_replace='Soon..', value= '1900-01-01', inplace=True)
juegos['release_date'].replace(to_replace='Beta测试已开启', value= '1900-01-01', inplace=True)
juegos['release_date'].replace(to_replace='Jun 2009', value= '2009-06-01', inplace=True)
juegos['release_date'].replace(to_replace='Oct 2010', value= '2010-10-01', inplace=True)
juegos['release_date'].replace(to_replace='Feb 2011', value= '2011-02-01', inplace=True)
juegos['release_date'].replace(to_replace='Aug 2014', value= '2014-08-01', inplace=True)
juegos['release_date'].replace(to_replace='Sep 2014', value= '2014-09-01', inplace=True)
juegos['release_date'].replace(to_replace='Apr 2015', value= '2015-04-01', inplace=True)
juegos['release_date'].replace(to_replace='Apr 2016', value= '2016-04-01', inplace=True)
juegos['release_date'].replace(to_replace='Jul 2016', value= '2016-07-01', inplace=True)
juegos['release_date'].replace(to_replace='June 2016', value= '2016-05-01', inplace=True)
juegos['release_date'].replace(to_replace='Coming Soon', value= '1900-01-01', inplace=True)
juegos['release_date'].replace(to_replace='Q2 2017', value= '2017-04-01', inplace=True)
juegos['release_date'].replace(to_replace='TBA', value= '1900-01-01', inplace=True)
juegos['release_date'].replace(to_replace= "When it's", value= '1900-01-01', inplace=True)
juegos['release_date'].replace(to_replace= "coming soon", value= '1900-01-01', inplace=True)

###############################
# esto podria haberse hecho de alguna manera mas eficiente.... para otra oportunidad...
###############################

Luego de estos remplazos, cambio el tipo de dato mediante datetime. Utilizo el parametro errors= 'coerce' para que los valores que no pueden transformarse o que dan error, devuelvan el valor NaT.

In [18]:
juegos['release_date']=pd.to_datetime(juegos['release_date'], errors= 'coerce', format= '%Y/%m/%d')

Ahora paso a tipo de dato str nuevamente y cambio nos 'NaT' por el mismo valor de fecha por defecto '1900-01-01' y luego vuelvo a cambiar al tipo datetime, quedando asi la columna limpia sin errores.


In [19]:
juegos['release_date']=juegos['release_date'].astype('str')
juegos['release_date'].replace(to_replace= 'NaT', value= '1900-01-01', inplace=True)

In [20]:
juegos['release_date']=pd.to_datetime(juegos['release_date'], errors= 'coerce', format= '%Y/%m/%d')

In [21]:
pd.isnull(juegos).sum()

publisher             0
genres             3283
app_name              0
url                   0
release_date          0
tags                163
discount_price    31910
reviews_url           2
specs               670
price              1377
early_access          0
id                    2
developer          3299
sentiment          7182
metascore         29458
dtype: int64

### 7-Creación de la columna Year. 

In [22]:
juegos.insert(loc=5, column='Year', value=2) # la creación de esta columna es necesaria para responder las consultas que plantea el enunciado y es importante para el modelo.
juegos['Year'] = juegos['release_date'].dt.strftime('%Y')

In [23]:
juegos

publisher                                             genres  \
0             Kotoshiro      [Action, Casual, Indie, Simulation, Strategy]   
1      Making Fun, Inc.               [Free to Play, Indie, RPG, Strategy]   
2          Poolians.com  [Casual, Free to Play, Indie, Simulation, Sports]   
3                  彼岸领域                        [Action, Adventure, Casual]   
4               no_info                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games              [Casual, Indie, Simulation, Strategy]   
32131            Sacada                          [Casual, Indie, Strategy]   
32132      Laush Studio                        [Indie, Racing, Simulation]   
32133          SIXNAILS                                    [Casual, Indie]   
32134           no_info                                                NaN   

                       app_name  \
0           Lost Summoner Kitty   
1                     Ironbound   
2       Real Pool 3D - Poolians   
3                       弹炸人2222   
4                 Log Challenge   
...                         ...   
32130            Colony On Mars   
32131  LOGistICAL: South Africa   
32132             Russian Roads   
32133       EXIT 2 - Directions   
32134               Maze Run VR   

                                                     url release_date  Year  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04  2018   
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04  2018   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24  2017   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07  2017   
4      http://store.steampowered.com/app/773570/Log_C...   1900-01-01  1900   
...                                                  ...          ...   ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04  2018   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04  2018   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04  2018   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02  2017   
32134  http://store.steampowered.com/app/681550/Maze_...   1900-01-01  1900   

                                                    tags  discount_price  \
0          [Strategy, Action, Indie, Casual, Simulation]            4.49   
1      [Free to Play, Strategy, Indie, RPG, Card Game...             NaN   
2      [Free to Play, Simulation, Sports, Casual, Ind...             NaN   
3                            [Action, Adventure, Casual]            0.83   
4                        [Action, Indie, Casual, Sports]            1.79   
...                                                  ...             ...   
32130              [Strategy, Indie, Casual, Simulation]            1.49   
32131                          [Strategy, Indie, Casual]            4.24   
32132                        [Indie, Simulation, Racing]            1.39   
32133  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...             NaN   
32134  [Early Access, Adventure, Indie, Action, Simul...             NaN   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                                                   specs         price  \

### 8- Columna tags.  

No se realiza transformacion sobre esta columna dada su composicion del tipo una lista por cada registro.

### 9- Columna 'discount price'  

In [24]:
juegos['discount_price']=juegos['discount_price'].replace(to_replace=np.nan , value=0)	

### 10- Columna 'reviews_url'

In [25]:

juegos['reviews_url']=juegos['reviews_url'].astype('str')

#pd.isnull(juegos).sum()
juegos['reviews_url'].replace(to_replace= 'nan', value= 'no_info', inplace=True)
#juegos.loc[juegos['reviews_url'] == 'no_info']


### 11- Columna 'specs'

Al igual que las otras columnas, no se realiza transformacion porque en cada uno de sus registros contiene una lista


### 12- Columna 'price'

Se remplazan valores no numéricos (por ej, 'Free To Play) con 0. De esta manera en un futuro para predecir el price del juego, se tomarán los valores > 0.

In [26]:

juegos['price'].replace(to_replace='Free To Play', value=pd.NA, inplace=True)
juegos['price'] = pd.to_numeric(juegos['price'], errors='coerce')
juegos['price'].fillna(0, inplace=True)
#juegos.info()
pd.isnull(juegos).sum()

publisher             0
genres             3283
app_name              0
url                   0
release_date          0
Year                  0
tags                163
discount_price        0
reviews_url           0
specs               670
price                 0
early_access          0
id                    2
developer          3299
sentiment          7182
metascore         29458
dtype: int64

### 13- Columna 'early_acces'


Esta columna no la voy a modificar.

### 14- Columna 'id'

In [27]:
#juegos
#juegos.loc[juegos['id'].isna()== True]
juegos['id']=juegos['id'].astype('str')

#pd.isnull(juegos).sum()
juegos['id'].replace(to_replace= 'nan', value= '0', inplace=True)
juegos['id'] = pd.to_numeric(juegos['id'], errors='coerce')
#juegos.info()


In [28]:
pd.isnull(juegos).sum()

publisher             0
genres             3283
app_name              0
url                   0
release_date          0
Year                  0
tags                163
discount_price        0
reviews_url           0
specs               670
price                 0
early_access          0
id                    0
developer          3299
sentiment          7182
metascore         29458
dtype: int64

### 15- Columna 'developer'

In [29]:
juegos['developer']=juegos['developer'].astype('str')
juegos['developer'].replace(to_replace= 'nan', value= 'no_info', inplace=True)
#pd.isnull(juegos).sum()


In [30]:
juegos

publisher                                             genres  \
0             Kotoshiro      [Action, Casual, Indie, Simulation, Strategy]   
1      Making Fun, Inc.               [Free to Play, Indie, RPG, Strategy]   
2          Poolians.com  [Casual, Free to Play, Indie, Simulation, Sports]   
3                  彼岸领域                        [Action, Adventure, Casual]   
4               no_info                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games              [Casual, Indie, Simulation, Strategy]   
32131            Sacada                          [Casual, Indie, Strategy]   
32132      Laush Studio                        [Indie, Racing, Simulation]   
32133          SIXNAILS                                    [Casual, Indie]   
32134           no_info                                                NaN   

                       app_name  \
0           Lost Summoner Kitty   
1                     Ironbound   
2       Real Pool 3D - Poolians   
3                       弹炸人2222   
4                 Log Challenge   
...                         ...   
32130            Colony On Mars   
32131  LOGistICAL: South Africa   
32132             Russian Roads   
32133       EXIT 2 - Directions   
32134               Maze Run VR   

                                                     url release_date  Year  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04  2018   
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04  2018   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24  2017   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07  2017   
4      http://store.steampowered.com/app/773570/Log_C...   1900-01-01  1900   
...                                                  ...          ...   ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04  2018   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04  2018   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04  2018   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02  2017   
32134  http://store.steampowered.com/app/681550/Maze_...   1900-01-01  1900   

                                                    tags  discount_price  \
0          [Strategy, Action, Indie, Casual, Simulation]            4.49   
1      [Free to Play, Strategy, Indie, RPG, Card Game...            0.00   
2      [Free to Play, Simulation, Sports, Casual, Ind...            0.00   
3                            [Action, Adventure, Casual]            0.83   
4                        [Action, Indie, Casual, Sports]            1.79   
...                                                  ...             ...   
32130              [Strategy, Indie, Casual, Simulation]            1.49   
32131                          [Strategy, Indie, Casual]            4.24   
32132                        [Indie, Simulation, Racing]            1.39   
32133  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...            0.00   
32134  [Early Access, Adventure, Indie, Action, Simul...            0.00   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                                                   specs  price  early_ac

### 16- Columna 'sentiment'


In [31]:
juegos['sentiment']=juegos['sentiment'].astype('str')
juegos['sentiment'].replace(to_replace= 'nan', value= 'no_info', inplace=True)
juegos

publisher                                             genres  \
0             Kotoshiro      [Action, Casual, Indie, Simulation, Strategy]   
1      Making Fun, Inc.               [Free to Play, Indie, RPG, Strategy]   
2          Poolians.com  [Casual, Free to Play, Indie, Simulation, Sports]   
3                  彼岸领域                        [Action, Adventure, Casual]   
4               no_info                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games              [Casual, Indie, Simulation, Strategy]   
32131            Sacada                          [Casual, Indie, Strategy]   
32132      Laush Studio                        [Indie, Racing, Simulation]   
32133          SIXNAILS                                    [Casual, Indie]   
32134           no_info                                                NaN   

                       app_name  \
0           Lost Summoner Kitty   
1                     Ironbound   
2       Real Pool 3D - Poolians   
3                       弹炸人2222   
4                 Log Challenge   
...                         ...   
32130            Colony On Mars   
32131  LOGistICAL: South Africa   
32132             Russian Roads   
32133       EXIT 2 - Directions   
32134               Maze Run VR   

                                                     url release_date  Year  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04  2018   
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04  2018   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24  2017   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07  2017   
4      http://store.steampowered.com/app/773570/Log_C...   1900-01-01  1900   
...                                                  ...          ...   ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04  2018   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04  2018   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04  2018   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02  2017   
32134  http://store.steampowered.com/app/681550/Maze_...   1900-01-01  1900   

                                                    tags  discount_price  \
0          [Strategy, Action, Indie, Casual, Simulation]            4.49   
1      [Free to Play, Strategy, Indie, RPG, Card Game...            0.00   
2      [Free to Play, Simulation, Sports, Casual, Ind...            0.00   
3                            [Action, Adventure, Casual]            0.83   
4                        [Action, Indie, Casual, Sports]            1.79   
...                                                  ...             ...   
32130              [Strategy, Indie, Casual, Simulation]            1.49   
32131                          [Strategy, Indie, Casual]            4.24   
32132                        [Indie, Simulation, Racing]            1.39   
32133  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...            0.00   
32134  [Early Access, Adventure, Indie, Action, Simul...            0.00   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                                                   specs  price  early_ac

### 17- Columna 'metascore'

Esta columna tiene 29458 valores nulos, entiendo que no aporta informacion util para sacar conclusiones. 


In [32]:
juegos['metascore']=juegos['metascore'].astype('str')
juegos['metascore'].replace(to_replace= 'nan', value= 'no_info', inplace=True)



In [33]:
juegos

publisher                                             genres  \
0             Kotoshiro      [Action, Casual, Indie, Simulation, Strategy]   
1      Making Fun, Inc.               [Free to Play, Indie, RPG, Strategy]   
2          Poolians.com  [Casual, Free to Play, Indie, Simulation, Sports]   
3                  彼岸领域                        [Action, Adventure, Casual]   
4               no_info                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games              [Casual, Indie, Simulation, Strategy]   
32131            Sacada                          [Casual, Indie, Strategy]   
32132      Laush Studio                        [Indie, Racing, Simulation]   
32133          SIXNAILS                                    [Casual, Indie]   
32134           no_info                                                NaN   

                       app_name  \
0           Lost Summoner Kitty   
1                     Ironbound   
2       Real Pool 3D - Poolians   
3                       弹炸人2222   
4                 Log Challenge   
...                         ...   
32130            Colony On Mars   
32131  LOGistICAL: South Africa   
32132             Russian Roads   
32133       EXIT 2 - Directions   
32134               Maze Run VR   

                                                     url release_date  Year  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04  2018   
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04  2018   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24  2017   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07  2017   
4      http://store.steampowered.com/app/773570/Log_C...   1900-01-01  1900   
...                                                  ...          ...   ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04  2018   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04  2018   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04  2018   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02  2017   
32134  http://store.steampowered.com/app/681550/Maze_...   1900-01-01  1900   

                                                    tags  discount_price  \
0          [Strategy, Action, Indie, Casual, Simulation]            4.49   
1      [Free to Play, Strategy, Indie, RPG, Card Game...            0.00   
2      [Free to Play, Simulation, Sports, Casual, Ind...            0.00   
3                            [Action, Adventure, Casual]            0.83   
4                        [Action, Indie, Casual, Sports]            1.79   
...                                                  ...             ...   
32130              [Strategy, Indie, Casual, Simulation]            1.49   
32131                          [Strategy, Indie, Casual]            4.24   
32132                        [Indie, Simulation, Racing]            1.39   
32133  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...            0.00   
32134  [Early Access, Adventure, Indie, Action, Simul...            0.00   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                                                   specs  price  early_ac

## Exportado de los datos:
Con el dataframe limpio, exporto en formato Json.

In [34]:
juegos.to_json('ETLJson.json', orient='columns', date_format='iso', date_unit='s')
#
#

In [35]:
juegos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32135 entries, 0 to 32134
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   publisher       32135 non-null  object        
 1   genres          28852 non-null  object        
 2   app_name        32135 non-null  object        
 3   url             32135 non-null  object        
 4   release_date    32135 non-null  datetime64[ns]
 5   Year            32135 non-null  object        
 6   tags            31972 non-null  object        
 7   discount_price  32135 non-null  float64       
 8   reviews_url     32135 non-null  object        
 9   specs           31465 non-null  object        
 10  price           32135 non-null  float64       
 11  early_access    32135 non-null  bool          
 12  id              32135 non-null  int64         
 13  developer       32135 non-null  object        
 14  sentiment       32135 non-null  object        
 15  me